docs/load_balancer.md

```diff
@@ -0,0 +1,278 @@
+# Load balancing & publicly accessible services
+
+This document describes how kubernetes services are made accessible outside the cluster.
+
+There are two main mechanisms:
+
+* The service can be exposed through a load-balancer managed by k8s.
+* k8s can also expose services without an associated load-balancer; the service could be reached via DNS SRV records,
+or by a load-balancer configured by a helper process, or even by a client using the k8s API directly for discovery.
+_(new functionality)_
+
+## Public ports
+
+To be exposed outside the k8s cluster, the service must be on known ports on externally reachable
+IPs.  For version 1 public services, k8s listens on _every_ external IP address on every minion, and assigns
+a _single_ port cluster-wide for each port on public services.
+
+## Important scenarios
+
+The names in bold are the names we will use in the API.
+
+1. **cluster** The service is reachable from within the cluster.  _(Existing functionality)_
+1. **loadbalancer** The service is exposed via a k8s-managed load-balancer.  _(Existing functionality)_
+1. **public** The service is exposed on a set of publicly-routable IP addresses, on a set of ports.  _(New functionality)_.
+The service endpoints would be discoverable via the k8s API, and helper processes could then expose that information
+ via DNS or configure external systems (e.g. physical load balancers)
+
+Expected future functionality:
+
+1. **namespace** The service is only accessible to pods belonging to the same namespace.  This is similar
+to _cluster_, but with additional firewall rules.
+1. **dmz** The service is accessible to pods that are in the DMZ.  This is similar to _cluster_,
+except that pods marked 'DMZ' are prevented from accessing _cluster_ services, and can only access services explicitly marked
+as dmz-reachable that are in their namespace.
+
+Currently these rules (and even the planned rules) nest very nicely (namespace < dmz < cluster < loadbalancer < public),
```

Comment:
You can have a load balancer without being public, but we couldn't know when to create that for you without making API changes that would only be supported by a small subset of cloud providers. You can also have a public IP without a load balancer, and we don't need any extra information in the API to support that, so public < loadbalancer would seem to make more sense in the near term.